### ISSUE TITLE
push.ha doesn't correctly count current failures if there is a previous failure

### ISSUE THREAD

**Issue by bitti** [voldemort/voldemort#472]
We have a 7 node store configured with `push.ha.max.node.failure=1`. Node 6 was failing for a while but fetch and swap was still working thanks to the `push.ha` configuration. But just as we got node 6 back we lost node 2. Still the fetch for the other 6 nodes was successful, but apparently `HandleFetchFailureResponse` doesn't count the number of failed nodes correctly:

```
2017-03-10 09:15:08,708 ERROR [pool-9-thread-1] voldemort.store.readonly.swapper.AdminStoreSwapper|tcp://redacted-node0.net:6666: tcp://redacted-node0.net:6666 : Error on Node redacted-node2.net:6666 [id 2] during push :
java.util.concurrent.ExecutionException: voldemort.store.UnreachableStoreException: Failure while checking out socket for redacted-node2.net:6666(vp1):
[...]
2017-03-10 09:15:08,710 INFO [pool-9-thread-1] voldemort.store.readonly.swapper.AdminStoreSwapper|tcp://redacted-node0.net:6666: tcp://redacted-node0.net:6666 : About to attempt: DisableStoreOnFailedNodeFailedFetchStrategy
2017-03-10 09:15:10,164 ERROR [pool-9-thread-1] voldemort.client.protocol.admin.AdminClient: Node iad1b-sn0003.iad.audsci.net:6666 [id 4] returned failed HandleFetchFailureResponse: We cannot use pushHighAvailability because it would bring the total number of nodes with disabled stores to more than 1...
2017-03-10 09:15:10,164 INFO [pool-9-thread-1] voldemort.store.readonly.swapper.AdminStoreSwapper|tcp://redacted-node0.net:6666: tcp://redacted-node0.net:6666 : Finished executing: DisableStoreOnFailedNodeFailedFetchStrategy; swapIsPossible: false
2017-03-10 09:15:10,164 INFO [pool-9-thread-1] voldemort.store.readonly.swapper.AdminStoreSwapper|tcp://redacted-node0.net:6666: tcp://redacted-node0.net:6666 : About to attempt: DeleteAllFailedFetchStrategy
2017-03-10 09:15:10,165 INFO [pool-9-thread-1] voldemort.store.readonly.swapper.DeleteAllFailedFetchStrategy: Deleting fetched data from node Node redacted-node0.net:6666 [id 0] in zone 0 partitionList:[0, 7, 14, 21, 28, 35, 42, 49, 56, 63, 70, 77, 84, 91, 98, 105, 112, 119, 126, 133, 140, 147, 154, 161, 168
2017-03-10 09:15:12,549 INFO [pool-9-thread-1] voldemort.store.readonly.swapper.DeleteAllFailedFetchStrategy: Deleting fetched data from node Node redacted-node1.net:6666 [id 1] in zone 0 partitionList:[1, 8, 15, 22, 29, 36, 43, 50, 57, 64, 71, 78, 85, 92, 99, 106, 113, 120, 127, 134, 141, 148, 155, 162, 169
2017-03-10 09:15:14,304 INFO [pool-9-thread-1] voldemort.store.readonly.swapper.DeleteAllFailedFetchStrategy: Deleting fetched data from node Node redacted-node3.net:6666 [id 3] in zone 0 partitionList:[3, 10, 17, 24, 31, 38, 45, 52, 59, 66, 73, 80, 87, 94, 101, 108, 115, 122, 129, 136, 143, 150, 157, 164, 1
2017-03-10 09:15:16,107 INFO [pool-9-thread-1] voldemort.store.readonly.swapper.DeleteAllFailedFetchStrategy: Deleting fetched data from node Node redacted-node4.net:6666 [id 4] in zone 0 partitionList:[4, 11, 18, 25, 32, 39, 46, 53, 60, 67, 74, 81, 88, 95, 102, 109, 116, 123, 130, 137, 144, 151, 158, 165, 1
2017-03-10 09:15:18,394 INFO [pool-9-thread-1] voldemort.store.readonly.swapper.DeleteAllFailedFetchStrategy: Deleting fetched data from node Node redacted-node5.net:6666 [id 5] in zone 0 partitionList:[5, 12, 19, 26, 33, 40, 47, 54, 61, 68, 75, 82, 89, 96, 103, 110, 117, 124, 131, 138, 145, 152, 159, 166, 1
2017-03-10 09:15:20,627 INFO [pool-9-thread-1] voldemort.store.readonly.swapper.DeleteAllFailedFetchStrategy: Deleting fetched data from node Node redacted-node6.net:6666 [id 6] in zone 0 partitionList:[6, 13, 20, 27, 34, 41, 48, 55, 62, 69, 76, 83, 90, 97, 104, 111, 118, 125, 132, 139, 146, 153, 160, 167, 1
2017-03-10 09:15:27,276 INFO [pool-9-thread-1] voldemort.store.readonly.swapper.AdminStoreSwapper|tcp://redacted-node0.net:6666: tcp://redacted-node0.net:6666 : Finished executing: DeleteAllFailedFetchStrategy; swapIsPossible: false
2017-03-10 09:15:27,277 ERROR [pool-9-thread-1] shell-job: Exception during push for cluster URL: tcp://redacted-node0.net:6666. Rethrowing exception.
```

I suspect the problem is due to the fact that this was the first time node 6 was back. This might be an edge case which isn't sufficiently tested.

**Comment by FelixGV** [comment]
Hi @bitti,

First of all, let me tell you a bit of background about BnP HA. We have been using BnP HA in production for about a year now, with several tweaks and fixes along the way. It "works" for us, but it has a lot of rough edges, and many known limitations.

It's great that you're trying it out, and hopefully that can pave the path forward to making it better outside of LinkedIn.

This is the crux of the situation:

> 2017-03-10 09:15:10,164 ERROR [pool-9-thread-1] voldemort.client.protocol.admin.AdminClient: Node iad1b-sn0003.iad.audsci.net:6666 [id 4] returned failed HandleFetchFailureResponse: We cannot use pushHighAvailability because it would bring the total number of nodes with disabled stores to more than 1...

This means the HA code thinks there is another node which is down, so it will not allow this one to go into "HA mode" as well. The way the HA code keeps track of that is via a HdfsFailedFetchLock implementation. The only such implementation available at this time is: HdfsFailedFetchLock, which can be configured via the push.ha.lock.implementation server config. With this implementation, the server code will store the state of the cluster in HDFS (which is not such a great place to store this, but it is the decision that was made in order to avoid pulling in ZK as a new dependency or some other homegrown solution). This state is stored in the path determined by the push.ha.lock.path server config. If you have left over directories there (from a previous invocation of BnP HA), then it can trip up future invocations. At this time, this state needs to be cleaned up manually during the recovery process.

Hope that helps.

I'm sure you'll have more questions. Feel free to keep them coming. I've got to run for now...

-F

**Comment by bitti** [comment]
Sure that helps! We still have the node-6 ha-lock files lying around on hdfs, so that's probably the issue. I try to make sure that we clean them up after each push. Or is there any worth to keep/archive them for debugging purposes?

Btw. is this documented somewhere? Or can we add documentation for this and some other push.ha aspects?

**Comment by FelixGV** [comment]
We have internal documentation which has been in flux. Now that the feature has somewhat stabilized, it would be good to have some public doc as well which would be more suited to the open-source. Alas, we don't have that yet, though I guess the original commit message for the feature could serve as a starting point: https://github.com/voldemort/voldemort/commit/c2db8fd9b0afe714f6f88908c769b66485d28825

Regarding your question at hand: the node-6 directory must be deleted after the node is properly recovered (which is a whole other can of worms, since it's not currently automated). The after-lock-release directory can be kept around for traceability, or can be cleared out periodically/systematically (it does not have any impact on the BnP HA behavior).

**Comment by bitti** [comment]
Yes I also started from that commit message. The thing is that this message seems currently be the only documentation available outside of linkedIn. 

Do you really have to do manual cleanups after a server fix? Why not just simply cleaning up the corresponding lock directory before attempting a push? What do you do for transient errors, e.g. network problems, when no manual fix was necessary to get the "node properly recovered"?

**Comment by FelixGV** [comment]
Hi @bitti,

Typically, when restoring a node, you would need to copy data onto the new healthy node. In some types of failure scenarios, you can copy it from the bad node, which is somewhat easy (glossing over some details here). If, however, the down node is completely unrecoverable, then you need to restore the data from peer nodes, which are partitioned differently. This is something we do here but is more of a headache.

Besides the above, there may also be concerns around data consistency. For example, if you resurrect a node back to a healthy state which contains some old data, this means that a fraction of the queries will be served by stale data. Again, this is something we take care of here at LinkedIn with a carefully executed runbook and some tooling, but I'm afraid the situation is not as robust in open-source.

We ought to release the runbook and the missing pieces of this story...

-F

**Comment by bitti** [comment]
Thanks for the information and I'm looking forward to the runbook and any documentation you could provide for the community.

You didn't answer though what you do for transient problems, i.e. when no manual intervention is required. I'm thinking about just automatically cleaning up all `{push.ha.lock.path}/{push.ha.cluster.id}/node-*/{push.store.name}` directories before pushing a specific store, to ensure that stale lockfiles won't let recovered nodes get basically "blacklisted". 

But it wonder if this would be the right way to go about it? I surely can't clean up just `{push.ha.lock.path}/{push.ha.cluster.id}` because there might be pushes of other stores in progress in parallel.

**Comment by FelixGV** [comment]
Yeah, so for transient node issues, the concern is again about data consistency. If you're ok with serving partially stale data coming from a previous push, then you can simply bring that node back online as is, delete the HDFS shared state directory, and let BnP refresh the cluster wide data on its next push.

**Comment by bitti** [comment]
So to be clear: a failed fetch during a successful push.ha will always set the node with the stale data in `OFFLINE_SERVER` state? And to bring this node back to `NORMAL_SERVER` it has to be restarted? Or is there some admin command for that? Or is it enough to clean the stale lockfiles?

But I don't understand why this should imply stale data. I'm ok with a server getting into `OFFLINE_SERVER` mode after a failed run. But if we get a successful fetch in a later push I of course would like to have it automatically back. If I understand your comment in c2db8fd correctly a push should even work in `OFFLINE_SERVER` mode when we use the admin port for pushing.

**Comment by bitti** [comment]
It seems it's not enough to clean up the ha lock files, but the server needs to be manually set back into `NORMAL` state by issuing a `vadmin.sh meta set offline=false -u tcp://some.node:6667`, or do I misunderstand something? Currently we are using server Version 1.10.14.

**Comment by FelixGV** [comment]
You are correct,

You also need to have a way of successfully restoring the data on the node that was down. That can be done with a new BnP job run, if you're willing to accept serving stale data in the mean time. Or it requires copying files from peer nodes, for which there isn't a very good solution available in the open-source at this time.

-F

**Comment by bitti** [comment]
I see. So a node in `OFFLINE_SERVER` will never fetch updated data? Shouldn't it be possible to enable the `readonly.fetch.enabled` flag even in offline mode to get a complete BnP run? It makes sense to block an offline server from delivering data when one wants to prevent serving outdated data, but why are updates are also prevented? (Note that we use the Admin port 6667 for pushes)

**Comment by FelixGV** [comment]
Hi,

That is what we'd like to explore. The original intent was to flip that flag and allow pushes to go into an offline node, but we have not tested this strategy so far, so it may have rough edges.

Exciting stuff ;) !

-F

**Comment by bitti** [comment]
I see. I might try it out when I encounter the problem next time. (Even though stale data isn't really a big problem for us.) On the other hand our nodes are running more stable now so it might take a while till I get the chance.

In the meantime I would love to close this ticket if you could push some documentation of the current behaviour? (Otherwise I could try to do it, but it surely wouldn't be as good as yours...)
